### PR TITLE
feat(sparq-mpc): BGW degree-reduction for chained multiplications (sq-dvuc)

### PR DESCRIPTION
**Pull request by jeswr**
> 🤖 _SPARQ agent_

## What & why (bead `sq-dvuc`)

`shamir::mul_shares_raw` produces a **degree-`2t`** product sharing and the crate
only ever OPENED it once (the equality test), so any computation deeper than a
single multiplication was impossible — the degree was never brought back to `t`.
This PR adds the missing **degree-reduction round** so secret-shared
multiplications can **chain**, which is the hard dependency for the
secure-threshold verdict (bead **sq-rrz4** builds on it).

## The protocol — BGW with a public recombination vector (chosen over DN double-sharings)

New `ShamirDealer::degree_reduce(shares_2t) -> Result<Vec<Share>, MpcError>`,
implementing the standard **BGW** reduction (Gennaro–Rabin–Rabin'98 simplification
of BGW'88) over the existing in-process party simulation:

1. The component-wise product `h_i = f(x_i)·g(x_i)` lies on a degree-`2t`
   polynomial `H` with `H(0) = a·b`. Recovering `H(0)` from its first `2t+1`
   points is a **fixed public linear map** — the Lagrange-at-0 weights
   `λ_1..λ_{2t+1}` (new RNG-free helper `lagrange_zero_weights`):
   `H(0) = Σ_i λ_i·h_i`.
2. Each of the `2t+1` recombination parties **re-shares** its own degree-`2t`
   share `h_i` under a **fresh, independent degree-`t`** polynomial (this dealer's
   masking RNG — a fresh OS-seeded ChaCha20 CSPRNG in production, sq-1vt). This is
   the one (simulated) communication round.
3. Every party locally applies the same public vector:
   `[a·b]_t = Σ_i λ_i·[h_i]_t`. The result is a fresh **degree-`t`** sharing whose
   secret is `Σ_i λ_i·H(x_i) = H(0) = a·b`.

**Why BGW-matrix not DN double-sharings:** the crate already has a single-dealer
re-sharing primitive (`ShamirDealer::share`) and exact Lagrange machinery, so the
public-recombination form is a direct, small reuse of what exists. DN
double-sharings (`[r]_t`, `[r]_{2t}` pairs) are the right call when you need O(n)
amortised communication across many gates in a *real* network; here, over an
in-process simulation with no wire-cost model, the simpler BGW-matrix reduction is
cleaner and provably equivalent for correctness/confidentiality. (DN remains an
acceptable future swap behind the same API.)

## Precondition — fail-closed (`n >= 2t+1`)

`degree_reduce` requires `n >= 2t+1` (the degree-`2t` polynomial must be
over-determined and the `2t+1` recombination points must exist). The
honest-majority constructor already fixes `t = ⌊(n−1)/2⌋` so every backend it
builds satisfies this; the check returns a **descriptive `MpcError::Protocol`**
(never a panic) if a short / non-canonical share vector is ever passed, and also
rejects a vector that is not the full `n`-party sharing.

## Security model — honest-majority / semi-honest (NOT malicious)

Unchanged trust assumptions from the rest of the backend: correct and
confidentiality-preserving when every party follows the protocol and at most `t`
collude. Each re-sharing draws fresh random masking coefficients, so any `≤ t`
parties' view of the sub-shares is independent of the reduced secret (the standard
BGW privacy argument). It is **NOT maliciously secure** — a deviating party can
feed an inconsistent re-sharing and, exactly as for the degree-`2t` equality open
at `n = 2t+1`, there is no in-protocol check here that detects it. Malicious
hardening (IT-MACs / verifiable resharing) is future work behind the SAME backend
and is **not claimed**. Vocabulary matches
`research/mpc-security-models-and-benchmarks.md` §3 (the "general Shamir
multiplication needs degree reduction" gap) and §6/§8 step 6.

## Tests (all four acceptance criteria, 178 pass)

1. `degree_reduce_round_trips_single_product` — reconstruct-after == reconstruct-before == plaintext product (incl. 0, 1, `P-1` edge values).
2. `two_multiplication_chain_a_b_c` — **the load-bearing test**: share a,b,c →
   mul → degree-reduce → mul → degree-reduce → reconstruct == plaintext `a·b·c`,
   across edge cases (zeros, identities, small primes, all-large `P-1`).
3. `degree_reduce_precondition_fails_closed` — `n < 2t+1` (and a wrong-`n` vector)
   return the descriptive error, **no panic**.
4. `reduced_sharing_is_genuinely_degree_t` — two different `t+1` subsets of the
   reduced shares both reconstruct the same secret at degree `t`, with a negative
   control proving `t+1` points of the *un-reduced* degree-`2t` product give the
   WRONG value (the reduction was real).

## Gate

- `cargo test -p sparq-mpc` → **178 passed; 0 failed**
- `cargo clippy -p sparq-mpc --all-targets -- -D warnings` → clean
- Touches only `crates/sparq-mpc/src/shamir.rs`. `[OPUS-4.8]` markers on new code.

## Unblocks

Removes the hard dependency for the £100k **secure-threshold** verdict (bead
**sq-rrz4**): chained multiplications enable secure comparison/threshold, products
of match conditions, and conjunctive hidden-pattern joins.

Refs bead **sq-dvuc**.

### CODE REVIEW

File: crates/sparq-mpc/src/shamir.rs [hunk 1]
```diff
@@ -338,10 +342,116 @@ impl ShamirDealer {
...
+                 got {} shares (honest-majority needs n >= 2t+1 to reduce a degree-2t product)",
+                2 * self.t + 1,
+                shares_2t.len()
+            )));
+        }
+        if shares_2t.len() != self.n {
+            return Err(MpcError::Protocol(format!(
+                "degree_reduce: expected the full {}-party sharing, got {} shares",
+                self.n,
+                shares_2t.len()
```
[the remark this answers is not in the document]
Comment (jeswr) [reply crates/sparq-mpc/src/shamir.rs:422]: > 🤖 _SPARQ agent_

Fixed in e33daf8. `degree_reduce` now performs an explicit runtime canonical-point check: after the `len`/`n` checks it verifies every input share is on `x = i+1` for `i = 0..n` (i.e. the full n-party sharing on `x = 1..=n` *in order*), failing closed with `MpcError::Protocol` otherwise. I chose the strict-canonical contract (it matches the docs and is the only ordering any in-process backend builds) over deriving Lagrange weights from arbitrary `x`. This guarantees `lagrange_zero_weights` only ever sees the canonical `x = 1..=2t+1` (distinct, nonzero), so it can neither panic in `Fp::inv` nor compute a recombination vector mismatched to the caller ordering. Added `degree_reduce_rejects_permuted_or_noncanonical_input` covering permuted order, off-lattice `x`, and `x = 0`; the 4 existing acceptance tests still pass. Doc updated to state the precondition is now enforced.

File: crates/sparq-mpc/src/shamir.rs [hunk 2]
```diff
@@ -338,10 +342,116 @@ impl ShamirDealer {
...
+                y: Fp::zero(),
+            })
+            .collect();
+        for (i, sub) in resharings.iter().enumerate() {
+            let lambda = lambdas[i];
+            for (out, sub_share) in reduced.iter_mut().zip(sub.iter()) {
+                debug_assert_eq!(out.x, sub_share.x, "resharing must use canonical points");
+                out.y = out.y.add(lambda.mul(sub_share.y));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jeswr) [reply crates/sparq-mpc/src/shamir.rs:494]: > 🤖 _SPARQ agent_

Fixed in e33daf8. The `debug_assert_eq!(out.x, sub_share.x, ...)` is replaced by a real `if out.x != sub_share.x { return Err(MpcError::Protocol(..)) }`, so the x-alignment invariant now holds in BOTH debug and release builds — a permuted-but-valid vector can no longer silently mis-accumulate when asserts are compiled out. Combined with the canonical-point precondition check (other thread), the by-position pairing of recombination weights / fresh sub-sharings against `share()`s canonical `x = 1..=n` output is now guaranteed correct. Proven by running the full suite under `--release`: `cargo test -p sparq-mpc --release` → 179 passed, including the new permuted/non-canonical rejection test (which would be a no-op under a debug-only assert in release).
